Adding parameters list and examples to specials submodule

Soumi7 opened the pull request:
<!-- Updated docstrings. Added parameter lists
and Examples section to functions in Special submodule of functions module -->

#### References to other Issues or PRs
<!-- Fixes #17887 -->


#### Brief description of what is fixed or changed

- Added parameters section to docstring jn_zeros function.
- Added parameters section to Bspline_basis, bspline_basis_set.
- Added parameters section to Interpolating_spline
- Added parameters section to Fdiff, eval (under DiracDelta)
- Added parameters section to Fdiff, eval (under Heaviside)
- Added parameters section to Fdiff, eval (under Heaviside)
- Added the example for under the text “Defining the li function via an integral:” in error_functions
- Added parameters section in Jacobi_normalized docstring
- Added examples in Ynm_c in spherical_harmonics
- Fixed and improved indentations for Parameters list
- Fixed trailing whitespaces

#### Other comments



#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

czgdp1807 commented:
Which issue is this PR fixing?

Soumi7 commented:
#17887

oscarbenjamin commented:
The PR should have a better name than "Branch2". The name should describe what the PR does.

Soumi7 commented:
I have updated the name

Soumi7 commented:
What about merging this PR? 
@asmeurer

oscarbenjamin commented on sympy/functions/special/delta_functions.py:
```diff
@@ -193,6 +198,12 @@ class is about to be instantiated and it returns either some simplified
+        Parameters
+        ==========
+
+        cls: DiracDelta class
```
We don't need to list the cls argument for a classmethod.

Soumi7 replied:
I will remove it then.

asmeurer commented on sympy/functions/special/error_functions.py:
```diff
@@ -1399,6 +1409,12 @@ class li(Function):
     1/log(z)
 
     Defining the ``li`` function via an integral:
+    >>> from sympy import integrate
```
I think you need an empty space above the example lines or it won't render correctly.

Soumi7 replied:
Yes, added it.

asmeurer commented:
You should build the docs to make sure everything rendered correctly, if you haven't already.

Sc0rpi0n101 commented on sympy/functions/special/delta_functions.py:
```diff
@@ -132,6 +132,11 @@ def fdiff(self, argindex=1):
+        Parameters
+        ==========
+
+        argindex : integer degree of derivative
```
```suggestion
        argindex : integer 
            degree of derivative
```

Sc0rpi0n101 commented on sympy/functions/special/delta_functions.py:
```diff
@@ -465,6 +475,11 @@ def fdiff(self, argindex=1):
+        Parameters
+        ==========
+
+        argindex : integer order of derivative
```
```suggestion
        argindex : integer 
            order of derivative
```

Sc0rpi0n101 commented on sympy/functions/special/delta_functions.py:
```diff
@@ -528,6 +543,13 @@ class is about to be instantiated and it returns either some simplified
+        Parameters
+        ==========
+
+        cls : Heaviside class
```
There's no need for this

Sc0rpi0n101 commented on sympy/functions/special/delta_functions.py:
```diff
@@ -528,6 +543,13 @@ class is about to be instantiated and it returns either some simplified
+        ==========
+
+        cls : Heaviside class
+        arg : argument passed my heaveside object
```
```suggestion
        arg : argument passed by heaveside object
```

Sc0rpi0n101 commented on sympy/functions/special/delta_functions.py:
```diff
@@ -193,6 +198,11 @@ class is about to be instantiated and it returns either some simplified
+        Parameters
+        ==========
+
+        arg : argument passed to DiracDelta
```
Add k too

Sc0rpi0n101 commented on sympy/functions/special/polynomials.py:
```diff
@@ -243,6 +243,13 @@ def jacobi_normalized(n, a, b, x):
     jacobi(n, a, b, x)/sqrt(2**(a + b + 1)*gamma(a + n + 1)*gamma(b + n + 1)/((a + b + 2*n + 1)*factorial(n)*gamma(a + b + n + 1)))
 
+    Parameters
+    ==========
+    n : integer degree of polynomial
```
```suggestion
    ==========

    n : integer degree of polynomial
```

Soumi7 replied:
Done

Soumi7 commented on sympy/functions/special/bessel.py:
```diff
@@ -984,6 +984,13 @@ def jn_zeros(n, k, method="sympy", dps=15):
+    Parameters
+    ==========
+
+    n : integer order of Bessel function
```
Made suggested changes.

Soumi7 commented on sympy/functions/special/bsplines.py:
```diff
@@ -254,6 +268,14 @@ def interpolating_spline(d, x, X, Y):
+    d : Degree of Bspline
+    x : symbol
+    X : list of integer values
+    Y : list of integer values
```
Done.

Soumi7 commented:
Have made the changes. #gsod

asmeurer commented on sympy/functions/special/bsplines.py:
```diff
@@ -137,6 +137,13 @@ def bspline_basis(d, knots, n, x):
+
+    d : degree of bspline
+    knots : list of knots points of bspline
+    n : $n$-th B-spline
```
You are missing the parameter `x` here

Soumi7 replied:
Added

asmeurer commented on sympy/functions/special/bsplines.py:
```diff
@@ -254,6 +268,17 @@ def interpolating_spline(d, x, X, Y):
+    ==========
+
+    d : integer
+    Degree of Bspline strictly greater than equal to one
```
Does this need to be indented to render properly?

Soumi7 replied:
Done!

asmeurer commented on sympy/functions/special/delta_functions.py:
```diff
@@ -132,6 +132,12 @@ def fdiff(self, argindex=1):
+        ==========
+
+        argindex : integer
+                degree of derivative
```
I'm not sure how strict Sphinx is over indentation but this should probably either line up with the line above or be indented 4 spaces.

Soumi7 replied:
I have indented it now.

asmeurer commented:
I'm unsure about the indentation formatting in the parameters section. Other than my comments here, everything else looks fine.

Soumi7 commented:
I have made all the changes.

Soumi7 commented:
#gsod

Soumi7 commented:
@asmeurer I think the formatting is fine now.

asmeurer commented on sympy/functions/special/delta_functions.py:
```diff
@@ -132,6 +132,12 @@ def fdiff(self, argindex=1):
+        ==========
+
+        argindex : integer
+        degree of derivative
```
Does this need to be indented?

Soumi7 replied:
Have fixed it.

Soumi7 commented:
Have fixed it.

Soumi7 commented:
GSoD
How about merging the PR? 
@asmeurer I have made all suggested changes.

oscarbenjamin commented on sympy/functions/special/bsplines.py:
```diff
@@ -254,6 +279,20 @@ def interpolating_spline(d, x, X, Y):
+    Y : list of strictly increasing integer values
+        list of Y coordinates through which the spline passes
+
+    x : symbol
```
Maybe the parameters should be in the same order as for the function

Soumi7 replied:
Done.

oscarbenjamin commented on sympy/functions/special/delta_functions.py:
```diff
@@ -528,6 +548,12 @@ class is about to be instantiated and it returns either some simplified
+        ==========
+
+        arg : argument passed by heaveside object
+        HO : boolean flag for HeaviSide Object is set to True
```
I guess it should be `Heaviside` (in both places)

Soumi7 replied:
Done.

oscarbenjamin commented:
In general this looks good. I hope it can be merged.

Sc0rpi0n101 commented on sympy/functions/special/polynomials.py:
```diff
@@ -1203,8 +1214,7 @@ class assoc_laguerre(OrthogonalPolynomial):
     alpha : Expr
-        Arbitrary expression. For ``alpha=0`` regular Laguerre
-        polynomials will be generated.
+        Arbitrary expression. For ``alpha=0`` regular Laguerre polynomials will be generated.
```
This line crosses 80 characters. You don't need to change this. 
It was done so that the code follows PEP-8(https://www.python.org/dev/peps/pep-0008/)

Soumi7 replied:
Okay sure

Sc0rpi0n101 commented on sympy/functions/special/bsplines.py:
```diff
@@ -137,6 +137,20 @@ def bspline_basis(d, knots, n, x):
         >>> f = lambdify(x, b0)
         >>> y = f(0.5)
 
+    Parameters
```
I'm not sure about this, but I think the Parameters would be better before examples.
It might be better to show the users what each parameter means, before showing them examples of how to use the function.

Soumi7 replied:
Even I think so. But from the beginning in all docs the parameters section has been added after Examples section. Should I change the whole thing?

Sc0rpi0n101 commented on sympy/functions/special/spherical_harmonics.py:
```diff
@@ -237,6 +237,29 @@ def Ynm_c(n, m, theta, phi):
+    >>> from sympy.abc import n,m
+    >>> theta = Symbol("theta")
+    >>> phi = Symbol("phi")
+
```
I'm not sure how good an idea these spaces are. 
Although, the examples look better in source, wouldn't it render as different shells?

Can you post some pictures of how the examples are rendered?
You can check https://docs.sympy.org/dev/documentation-style-guide.html#build-the-documentation for building the documentation locally.

Soumi7 replied:
I am building the docs and checking locally everytime before adding commits to this PR.

Soumi7 replied:
![Screenshot from 2020-06-03 12-52-51](https://user-images.githubusercontent.com/51290447/83607736-5dc51d80-a599-11ea-8c86-1324d7e37aae.png)

Soumi7 replied:
This is how it is rendering each example

Sc0rpi0n101 replied:
Yeah, the examples are rendering as different shells. 
I don't think the tiny bit of source readability is worth rendering the examples as different shells.

Soumi7 commented:
Made changes.

Sc0rpi0n101 commented on sympy/functions/special/spherical_harmonics.py:
```diff
@@ -237,6 +237,29 @@ def Ynm_c(n, m, theta, phi):
+    >>> from sympy.abc import n,m
+    >>> theta = Symbol("theta")
+    >>> phi = Symbol("phi")
+
```
You should remove this line

Soumi7 replied:
Done

Sc0rpi0n101 commented on sympy/functions/special/spherical_harmonics.py:
```diff
@@ -237,6 +237,29 @@ def Ynm_c(n, m, theta, phi):
+
+    >>> Ynm_c(n, m, theta, phi)
+    (-1)**(2*m)*exp(-2*I*m*phi)*Ynm(n, m, theta, phi)
+
```
and this line

Soumi7 replied:
Done

Sc0rpi0n101 commented on sympy/functions/special/spherical_harmonics.py:
```diff
@@ -237,6 +237,29 @@ def Ynm_c(n, m, theta, phi):
+
+    >>> simplify(Ynm_c(0, 0, theta, phi).expand(func=True))
+    1/(2*sqrt(pi))
+
```
and this one

Soumi7 replied:
Done

Sc0rpi0n101 commented:
Other than my one comment and the position of Parameters, which I think should come before the Examples instead of after. Everything else looks good to me.

Soumi7 commented:
Here is how the examples are being rendered :

![Screenshot from 2020-06-06 20-29-38](https://user-images.githubusercontent.com/51290447/83947488-d98ec680-a834-11ea-9400-858df7ebf827.png)

Soumi7 commented:
Regarding the parameter, @Sc0rpi0n101 what do you recommend I do? As for that existing format will change. For all functions.

Sc0rpi0n101 reviewed (approved):
The position of the Parameter section is a larger discussion than this PR as it affects the whole documentation.
 There is a similar discussion on the mailing list at https://groups.google.com/forum/#!topic/sympy/bXwxVEkbuLk

Everything else looks good to me. 
Can you squash your commits into one with a descriptive message? 
You can check https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages for reference.

Soumi7 commented:
@Sc0rpi0n101  I have squashed all the commits into one with a description now.

Sc0rpi0n101 commented:
This PR still has 20 commits. Can you squash all of them, and just have 1 or 2?

Soumi7 committed:
Updated docs : Parameters List and examples in Specials Submodule:

- Added parameters section in jn_zeros docstring

- Added parameters section to bspline_basis and bspline_basis_set

- Added Parameters list for Interpolating_spline in bsplines.py

- Added parameters list to Fdiff and eval in delta_functions.py

- Added parameters list for fdiff and eval under Heaviside functions in delta_functions.py

- Added example for the text "Defining the ```li ``` function via an integral in error_functions.py

- Added parameters section in jacobi_normalized in polynomials.py

- Added examples section for ynm_c

- Changes to docstring

- Updated docstrings

- Updated docstrings and whitespaces

- Removed the cls argument in delta_functions.py

- Made changes to the parameter lists in the mentioned function is suggestions.

- Fixed traniling whitespaces.

- Improved indentations

- Fixed indentations in deltafunctions.py

- Resolved comments.

- Rendering examples in spherical_harmonics.py

Soumi7 commented:
Is this fine now? @Sc0rpi0n101

Sc0rpi0n101 commented:
Great. 
Unless anyone has any more reviews, I'll merge it.

Soumi7 commented:
Yes !

Sc0rpi0n101 commented:
There's not a specific no of reviewers that are needed for a pull request. Just the changes need to be reviewed adequately. 

Merging this now, thank you for your contribution.
